### PR TITLE
Staging persistence layer

### PR DESCRIPTION
**Pull request by subodh1810**
Issue : https://github.com/airbytehq/airbyte-internal-issues/issues/426

TODO(Subodh) : Update the PR description

### CODE REVIEW

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/SecretsRepositoryReader.java
Comment (sherifnada) [inline airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/SecretsRepositoryReader.java:80]: Why are we exporting staging configs? I know it's not intended for use in OSS, but let's say we release a paid version of Airbyte self-hosted where this feature is enabled. Then we probably want to permission what can be dumped individually, rather than allow the user to dump the whole instnace. 

@jrhizor does the platform already have a stance on this? My vague feeling is we shouldn't dump at all here but would love feedback from you

How do we want to handle secrets which are really only meant for use in Cloud/paid?

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/split_secrets/SecretCoordinateToPayload.java
Comment (sherifnada) [inline airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/split_secrets/SecretCoordinateToPayload.java:17]: nit: can the constructor be the first declare method in the class?

Comment (sherifnada) [reply airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/split_secrets/SecretCoordinateToPayload.java:17]: actually can't this just be declared as 

```
public record SecretCoordinateToPayload(Secret Coordinate secretCoordinate, String payload, JsonNode secretCoordinateForDB) {}
```
and java would fill in the rest?

File: airbyte-config/persistence/src/test/java/io/airbyte/config/persistence/SecretsRepositoryWriterTest.java
Comment (sherifnada) [inline airbyte-config/persistence/src/test/java/io/airbyte/config/persistence/SecretsRepositoryWriterTest.java:223]: you can use multi-line strings (I believe even on a single line) to make this more readable

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/split_secrets/MemorySecretPersistence.java
Comment (sherifnada) [inline airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/split_secrets/MemorySecretPersistence.java:30]: why make this unmodifiable? I am a little scared of these classes because their signature is the same as a map but it can break at runtime

File: airbyte-api/src/main/openapi/config.yaml
Comment (jrhizor) [reply airbyte-api/src/main/openapi/config.yaml:2584]: Should the staging configuration have an id it can return? And the ability to get a specific configuration object from the api?

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/split_secrets/SecretsHelpers.java
Comment (jrhizor) [inline airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/split_secrets/SecretsHelpers.java:454]: This probably shouldn't be tied to the destination definition id. The workspace scoping above was to identify the "owner" of the secret. 

So it could be either `actor_definition` or `staging_config` too.

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/split_secrets/SecretsHelpers.java
Comment (jrhizor) [inline airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/split_secrets/SecretsHelpers.java:464]: nit: I'd prefer `hydrate` to `decrypt` since it's loading from an external store, not actually performing decryption.

**Review by jrhizor — CHANGES_REQUESTED** [review]

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/split_secrets/SecretsHelpers.java
Comment (girarda) [reply airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/split_secrets/SecretsHelpers.java:417]: probably because of https://github.com/airbytehq/airbyte/blob/master/airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/split_secrets/SecretCoordinate.java#L14-L17
> We use "full coordinate" to refer to a string reference that includes both the coordinate base and version-specific information

Comment (sherifnada) [reply airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/split_secrets/SecretsHelpers.java:417]: got it, thanks for the explanation. resolvign

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java
Comment (benmoriceau) [inline airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java:771]: instead of selecting `asterisk` can we select the 4 columns that are needed?

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/split_secrets/MemorySecretPersistence.java
Comment (subodh1810) [reply airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/split_secrets/MemorySecretPersistence.java:30]: @sherifnada the class `MemorySecretPersistence` is only being used in tests mainly to assert that the correct secret would be written to the persistence. The reason I convert it to unmodifiableMap is to safe guard it from being mutated in any way outside this class. This just gives me more confidence on the assertions. If you feel that we should not use unmodifiableMap, I can change it

Comment (girarda) [reply airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/split_secrets/MemorySecretPersistence.java:30]: should we move this class to a test package to ensure we don't accidentally use it in production?

**Review by girarda — APPROVED** [review]
left a couple of questions, but this looks good to me.

File: airbyte-api/src/main/openapi/config.yaml
Comment (michel-tricot) [inline airbyte-api/src/main/openapi/config.yaml:941]: is it a configuration binding or a configuration override?

File: airbyte-server/src/main/java/io/airbyte/server/handlers/ActorConfigurationBindingHandler.java
Comment (sherifnada) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/ActorConfigurationBindingHandler.java:13]: we should make a unit test for this

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/ConfigRepository.java
Comment (sherifnada) [inline airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/ConfigRepository.java:743]: nit: can you add a comment to both methods explaining the `NoSecrets` part?

Comment (sherifnada) [reply airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/ConfigRepository.java:743]: and rename it to `WithoutSecrets` for consistency with SecretsRepositoryReader?

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/SecretsRepositoryReader.java
Comment (sherifnada) [inline airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/SecretsRepositoryReader.java:100]: nit: `actorDefinitionID`

File: airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/split_secrets/MemorySecretPersistence.java
Comment (sherifnada) [reply airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/split_secrets/MemorySecretPersistence.java:30]: I like alex's suggestion here

**Review by sherifnada — APPROVED** [review]
LGTM modulo alex's/my comments. @jrhizor it may good for you to take a look too

**Comment by marcosmarxm** [comment]
Hello :wave:, first thank you for this amazing contribution. 

We really appreciate the effort you've made to improve the project.
We ask you patience for the code review. Last month our team was focused on Hacktoberfest event and that probably left some PR without the proper feedback. And this week, due to the Thanksgiving US Holiday, most our team is out of office with their families. Another important piece of information why code won't be merge this week is: as a safety measure the core team has decided to freeze merging code to main branch to keep the release stable. Next week we'll return to you with the proper code review and update the status of your contribution.

If you have any questions feel free to send me a message in Slack!
Thanks!

**Pull request closed without merging**
